subsystems/networking/bluetooth.go

```diff
@@ -0,0 +1,630 @@
+	}
+	return nil
+}
+
+/** Helper methods for low-level system calls and read/write requests to/from bluetooth characteristics **/
+
+func (bsl *bluetoothServiceLinux) readSsid() (string, error) {
+	if bsl.characteristicSsid == nil {
+		return "", errors.New("characteristic ssid is nil")
+	}
+
+	bsl.characteristicSsid.mu.Lock()
+	defer bsl.characteristicSsid.mu.Unlock()
+
+	if bsl.characteristicSsid.currentValue == nil {
+		return "", newEmptyBluetoothCharacteristicError("ssid")
+	}
+	return *bsl.characteristicSsid.currentValue, nil
+}
+
+func (bsl *bluetoothServiceLinux) readPsk() (string, error) {
+	if bsl.characteristicPsk == nil {
+		return "", errors.New("characteristic psk is nil")
+	}
+
+	bsl.characteristicPsk.mu.Lock()
+	defer bsl.characteristicPsk.mu.Unlock()
+
+	if bsl.characteristicPsk.currentValue == nil {
+		return "", newEmptyBluetoothCharacteristicError("psk")
+	}
+	return *bsl.characteristicPsk.currentValue, nil
+}
+
+func (bsl *bluetoothServiceLinux) readRobotPartKeyID() (string, error) {
+	if bsl.characteristicRobotPartKeyID == nil {
+		return "", errors.New("characteristic robot part key ID is nil")
+	}
+
+	bsl.characteristicRobotPartKeyID.mu.Lock()
+	defer bsl.characteristicRobotPartKeyID.mu.Unlock()
+
+	if bsl.characteristicRobotPartKeyID.currentValue == nil {
+		return "", newEmptyBluetoothCharacteristicError("robot part key ID")
+	}
+	return *bsl.characteristicRobotPartKeyID.currentValue, nil
+}
+
+func (bsl *bluetoothServiceLinux) readRobotPartKey() (string, error) {
+	if bsl.characteristicRobotPartKey == nil {
+		return "", errors.New("characteristic robot part key is nil")
+	}
+
+	bsl.characteristicRobotPartKey.mu.Lock()
+	defer bsl.characteristicRobotPartKey.mu.Unlock()
+
+	if bsl.characteristicRobotPartKey.currentValue == nil {
+		return "", newEmptyBluetoothCharacteristicError("robot part key")
+	}
+	return *bsl.characteristicRobotPartKey.currentValue, nil
+}
+
+/** Custom error type and miscellaneous utils that are helpful for managing low-level bluetooth on Linux **/
+
+// emptyBluetoothCharacteristicError represents the error which is raised when we attempt to read from an empty BLE characteristic.
+type emptyBluetoothCharacteristicError struct {
+	missingValue string
+}
+
+func (e *emptyBluetoothCharacteristicError) Error() string {
+	return fmt.Sprintf("no value has been written to BLE characteristic for %s", e.missingValue)
+}
+
+func newEmptyBluetoothCharacteristicError(missingValue string) error {
+	return &emptyBluetoothCharacteristicError{
+		missingValue: missingValue,
+	}
+}
+
+const (
+	BluezDBusService  = "org.bluez"
+	BluezAgentPath    = "/custom/agent"
+	BluezAgentManager = "org.bluez.AgentManager1"
+	BluezAgent        = "org.bluez.Agent1"
+)
+
+// checkOS verifies the system is running a Linux distribution.
+func checkOS() error {
+	if runtime.GOOS != "linux" {
+		return fmt.Errorf("this program requires Linux, detected: %s", runtime.GOOS)
+	}
+	return nil
+}
```

Comment:
This likely shouldn't exist. Use build flags to guard code that simply can't run elsewhere at compile time, not checking at runtime.